### ISSUE TITLE
Coding team tasks for the rest of spring, 2017

### ISSUE THREAD

**Issue by kdahlquist** [kdahlquist/GRNmap#314]
Tasks for @trixr4kdz, @cazinge, and @jtorre39:
- [x] create tests for LSE function #313 (priority 0)
- [x] meet up with @bengfitzpatrick to look at #310, begin converting LSE to new data structure (priority 0.5)
- [x] work on #311, converting existing test cases to the new data structure (priority 0.5)

Tasks for @kdahlquist:
- [ ] #323, different results on different computers
- [ ] review requested for 0% release #312, waiting on resolution of #323.

**Comment by cazinge** [comment]
Update for week:
* There's a draft for the 0% release (v1.6) for the current week up on GRNmap's releases page. Approval should be given during this week's meeting before we publish.
* We've revised the draft and seen @kdahlquist's comments; upon @dondi and @bengfitzpatrick's review we will submit to the symposium.
* We've partially reviewed the changes that we'll need to make in MSETest, as the file should be broken into multiple smaller unit tests from the large functional one that it is now. In theory, however, it's existing state should be able to be used as one large test just to ensure that LSE is working correctly; then we would iterate upon its tests and improve its quality after converting to the new structure. @bengfitzpatrick's feedback at this week's meeting would be appreciated. (Side note: runGRNstructSimulation does not seem like its being used, appearing within some commented out code. Should we remove the file and clean up the comments in callTests?)

**Comment by dondi** [comment]
Immediate priority is #301: remove strikeouts and make a final reading before submitting.

**Comment by kdahlquist** [comment]
I just want to record that MSE is not the same as LSE as discussed in the reading.  

It would be a good idea to review the Dahlquist et al. 2015 paper to understand the math behind the LSE.

**Comment by trixr4kdz** [comment]
For this week's session, we determined which variables need to be tested to ensure that the LSE function works as intended. From this, I've created templates for testing the lse and gLSE routines so that we only need to worry about figuring out what the expected values are for those routines.

**Comment by kdahlquist** [comment]
Task list has been updated for this week; see issue for updates.

**Comment by trixr4kdz** [comment]
For creating the LSE tests, see the comments on #313. 

Tl;dr: Testing the LSE function would require assuming that the current output is correct so that when we implement the compressMissingData function, the current output should not be changed. However, what this would mean is probably generating a new set of output for different use cases just like the "sixteen_tests" excel files, which are already different from the current output of the code. @cazinge suggested working in parallel in implementing the new data structure.

**Comment by kdahlquist** [comment]
Task list remains unchanged.  I noted at the meeting if you are planning to take off early next Friday for Spring Break, then you need to arrange to do your research hours earlier in the week.

**Comment by bengfitzpatrick** [comment]
@cazinge @trixr4kdz @jtorre39  please let me know when you're up-and-running in the UH lab.

**Comment by jtorre39** [comment]
@bengfitzpatrick We're set to go!

**Comment by jtorre39** [comment]
@bengfitzpatrick stopped by during our session and gave us specifics on how to approach our tests. Here is the result from our discussion. 
![Fitzpatrick's Whiteboard](https://cloud.githubusercontent.com/assets/16827894/23319995/c4e3a726-fa8d-11e6-94b9-d7ab4ad59393.jpg)


![Young Eddie's Whiteboard](https://cloud.githubusercontent.com/assets/16827894/23319966/9fc90422-fa8d-11e6-8aa9-38d5d0916b32.jpg)

**Comment by trixr4kdz** [comment]
For this week's meeting, @bengfitzpatrick helped us map out what needed to be in the gLSE tests. More specifically, two of our test cases were taken from:
- [3-gene, 3-edges](https://github.com/kdahlquist/GRNmap/blob/beta/test_files/estimation_tests/3-genes_3-edges_artificial-data_Sigmoid_estimation_1.xlsx)
- [4-gene, 6-edges](https://github.com/kdahlquist/GRNmap/blob/beta/test_files/estimation_tests/4-genes_6-edges_artificial-data_Sigmoid_estimation_output_archived.xlsx) 
Note that in both cases, the L = 0.

**Comment by trixr4kdz** [comment]
The next steps include working on more test cases that iterate through the test cases we already made and working out the answers by hand.

**Comment by trixr4kdz** [comment]
For reference, this is the equation that we will use for determining what the L would be based on our changes to the test data:

![16935544_1822496211346048_1130146500_o](https://cloud.githubusercontent.com/assets/8975300/23324795/082fdc54-faa5-11e6-9647-a98b6a47d3e8.jpg)

**Comment by dondi** [comment]
Main work at the meeting #313, with #310 and #311 to follow after #313.

**Comment by trixr4kdz** [comment]
I wrote some comments on new tests for issue #313. 

I have now started working on #310 on integrating the new data structure since the new tests for gLSE have been merged (PR #339).

**Comment by trixr4kdz** [comment]
Minor question, how does the SSE calculation change with the new data structure?

**Comment by kdahlquist** [comment]
microData will be changed to expressionData with the fields ```raw```, ```compressed```, ```strain```, ```avg```, ```deletion```, ```t```, `stdev`

`microData(index).data` in https://github.com/kdahlquist/GRNmap/blob/48da7b9d78a4640a38b20a64220967f489501706/matlab/readInputSheet.m#L61 becomes `expressionData(index).raw`

`expressionData(index).data` in https://github.com/kdahlquist/GRNmap/blob/beta/matlab/compressMissingData.m#L4 becomes `expressionData(index).compressed`

**Comment by jtorre39** [comment]
Courtesy of @bengfitzpatrick, a break down of the loop within GLSE
![2017-04-05 12 18 48](https://cloud.githubusercontent.com/assets/16827894/24722916/56a01c68-19fa-11e7-8c77-c9a3be18aaf5.jpg)

**Comment by kdahlquist** [comment]
I just wanted to note that at the meeting today, we advised this course of action:

1. Make sure the current LSE tests with the current data structure is functional (I believe that @jtorre39 confirmed that this was the case.
2. Make the modifications to the LSE code.
3. Run the test with the complete (no missing) data test used on the original LSE code and verify that you get the same results.
4. Write the additional tests needed to capture all the missing data cases that we previously listed.

**Comment by trixr4kdz** [comment]
For the points mentioned above by @kdahlquist:
- Finally finished modifying the gLSE function 
- I ran the tests and they are still passing (woot!)
- We just need to make more tests for item 4

**Comment by trixr4kdz** [comment]
So I ran the ALL the tests this time, even MSETest (which still takes ~30-ish minutes for just that 1 test unfortunately) and they all pass.

One thing that I observed when I ran the code with the `sixteen_tests` Excel files, however, is that the precision error we had with timepoint `1.2` came back. As a result, GRNmap will think that the third replicate for timepoint `1.2` is a different timepoint (i.e., we have 3 replicates for t=0.4, 3 reps for t=0.8, 2 reps for t=1.2, 1 rep for t=1.2...0001, and 3 reps for t=1.6)

**Comment by kdahlquist** [comment]
Closing out the work of last semester!
